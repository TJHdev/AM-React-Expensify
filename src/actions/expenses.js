import uuid from 'uuid';
import database from '../firebase/firebase.js'

// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// component calls action generator
// action generator returns function
// component dispatches object (?)
// function runs (has the ability to dispatch other actions and do whatever it wants)

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense: expense
})

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    
    const expense = { description, note, amount, createdAt };

    return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id: id
});

export const startRemoveExpense = ({ id }) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).set(null).then(() => {
      dispatch(removeExpense({ id }));
    });
  }
};

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id: id,
  updates
});

export const startEditExpense = (id, updates = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    }); 
  };
};

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses: expenses
});

export const startSetExpenses = (expensesData = []) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
      const expenses = [];
   
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })
   
      dispatch(setExpenses(expenses));
    });
  }
};



 
