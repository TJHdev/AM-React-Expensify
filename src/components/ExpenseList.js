import React from 'react';
import { connect } from 'react-redux';

const ExpenseList = (props) => (
  <div>
    <h1>ExpenseList</h1>
    {props.filters.text}
    {props.expenses}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expense: state.expenses,
    filters: state.filters
  };
};

export default connect(mapStateToProps)(ExpenseList);