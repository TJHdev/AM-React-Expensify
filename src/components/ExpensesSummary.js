import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = ({ expenseCountVisible, expensesTotalVisible, expenseCountHidden, expensesTotalHidden }) => {
  let pluralExpenses = expenseCountVisible === 1 ? '' : 's';
  let pluralHiddenExpenses = expenseCountHidden === 1 ? '' : 's';
  // let expensesTotal = selectExpensesTotal(props.expenses);
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{expenseCountVisible}</span> expense{pluralExpenses} 
          &nbsp;totalling <span>{numeral(expensesTotalVisible / 100).format('$0,0.00')}</span>
        </h1>
        <h1 className="page-header__title">
          Hiding <span>{expenseCountHidden}</span> expense{pluralHiddenExpenses} 
          &nbsp;totalling <span>{numeral(expensesTotalHidden / 100).format('$0,0.00')}</span>
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">Add Expense</Link>
        </div>
      </div>
    </div>
  );
}

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  const expensesTotalVisible = selectExpensesTotal(visibleExpenses);

  return {
    expenseCountVisible: visibleExpenses.length,
    expenseCountHidden: state.expenses.length - visibleExpenses.length,
    expensesTotalVisible: expensesTotalVisible,
    expensesTotalHidden: selectExpensesTotal(state.expenses) - expensesTotalVisible
  };
};

export default connect(mapStateToProps)(ExpensesSummary);





