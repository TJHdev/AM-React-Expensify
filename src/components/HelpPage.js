import React from 'react';

const HelpPage = () => (
  <div>
    This is from my edit HelpPage
  </div>
);

export default HelpPage;