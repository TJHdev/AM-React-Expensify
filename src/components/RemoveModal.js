import React from 'react';
import Modal from 'react-modal';

const RemoveModal = (props) => (
  <Modal
    isOpen={Boolean(props.confirmRemoveExpense)}
    onRequestClose={props.handleCloseRemoveModal}
    contentLabel="Confirm Remove Expense"
    closeTimeoutMS={200}
    className="modal"
    ariaHideApp={false}
  >
    <h3 className="modal__title">Confirm Delete Expense</h3>
    <div className="modal__container">
      <button className="button" onClick={props.onRemove}>Delete</button>
      <button className="button test" onClick={props.handleCloseRemoveModal}>No</button>
    </div>
  </Modal>
);

export default RemoveModal;