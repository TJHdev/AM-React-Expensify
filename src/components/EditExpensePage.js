import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

const EditExpensePage = (props) => {
  return  (
    <div>
      <h1>Edit expense {props.match.params.id}</h1>
      <ExpenseForm 
        expense={props.expense}
        onSubmit={(expense) => {
          props.dispatch(editExpense(props.expense.id, expense));
          props.history.push('/');
        }}
      />
      <button onClick={(e) => {
        console.log(props.match.params.id)
        props.dispatch(removeExpense({ id: props.expense.id }));
        props.history.push('/');
      }}>Remove</button>
    </div>
  );
};

const mapStateToProps = (state, props) => { // can pass both state and props
  return {
    expense: state.expenses.find((expense) => {
      return expense.id === props.match.params.id;
    })
  };
};

export default connect(mapStateToProps)(EditExpensePage);