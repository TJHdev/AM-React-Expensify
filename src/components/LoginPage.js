import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLoginGoogle, startLoginGithub, startLoginFacebook }) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">Expensify App</h1>
      <p>It's time to get your expenses under control.</p>
      <button className="button button--google" onClick={startLoginGoogle} >Login with Google</button>
      <button className="button" onClick={startLoginGithub} >Login with Github</button>
    </div>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLoginGoogle: () => dispatch(startLogin('google')),
    startLoginGithub: () => dispatch(startLogin('github'))
    // startLoginFacebook: () => dispatch(startLogin('facebook'))
  });

export default connect(undefined, mapDispatchToProps)(LoginPage);

// <button className="button" onClick={startLoginFacebook} >Login with Facebook</button>