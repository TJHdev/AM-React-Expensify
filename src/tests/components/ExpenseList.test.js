import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses'; 
// import ExpenseListItem from '../../components/ExpenseListItem';

test('should render ExpenseList with expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
}); // snapshot with some expenses

test('should render ExpenseList with empty message', () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});

// Create test file
// Grab imports
// Render ExpenseListItem
// Create snapshot
