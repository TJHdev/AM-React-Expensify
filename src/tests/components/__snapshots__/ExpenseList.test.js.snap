// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div>
  <h1>
    ExpenseList
  </h1>
  <p>
    There are no expenses
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <h1>
    ExpenseList
  </h1>
  <ExpenseListItem
    amount={19500}
    createdAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={20500}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={30500}
    createdAt={345600000}
    description="Credit Card"
    id="3"
    key="3"
    note=""
  />
</div>
`;
